### PR TITLE
ParkAPI integration

### PR DESCRIPTION
**Pull request by leonardehrenfried**
This PR adds an updater that dynamically adds car park using the ParkAPI format.

By default it completely ignores the available spaces in the car park and adds an alert instead.

However, you can also add take them into account and exclude full ones from the search result by setting the request parameter `useCarParkAvailabilityInformation = true`.

### CODE REVIEW

File: src/main/java/org/opentripplanner/routing/edgetype/ParkAndRideEdge.java [hunk 1]
```diff
@@ -66,6 +63,11 @@ public State traverse(State s0) {
             if (s0.isCarParked()) {
                 throw new IllegalStateException("Can't drive 2 cars");
             }
+            if(request.useCarParkAvailabilityInformation
```
Comment (leonardehrenfried) [inline src/main/java/org/opentripplanner/routing/edgetype/ParkAndRideEdge.java:66]: Just noticed that this ought to be moved up. Will change it.

File: src/main/java/org/opentripplanner/routing/car_park/CarPark.java [hunk 2]
```diff
@@ -75,4 +83,9 @@ public int hashCode() {
     public String toString () {
         return String.format(Locale.US, "Car park %s at %.6f, %.6f", name, y, x);
     }
+
+    public boolean hasFewSpacesAvailable() {
+        var percentFree = ((float) spacesAvailable / maxCapacity);
+        return !(Double.isNaN(percentFree)) && percentFree < 0.1;
```
Comment (hbruch) [inline src/main/java/org/opentripplanner/routing/car_park/CarPark.java:89]: We should have limits for absolute availability numbers as well, I think. For parkings with less than ten spaces, we should give an alert if only one space is available. More than 20 or 30 parkings IMHO is sufficient if the parking has more than 200 maxCapacity

**Review by hbruch — CHANGES_REQUESTED** [review]

File: src/main/java/org/opentripplanner/updater/car_park/ParkApiDataSource.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Note that the JSON being passed in here is for configuration of the OTP component, it's completely separate
+     * from the JSON coming in from the update source.
+     */
+    @Override
+    public void configure (Graph graph, JsonNode jsonNode) {
+        super.configure(graph, jsonNode);
+        String url = jsonNode.path("url").asText(); // path() returns MissingNode not null.
+        if (url != null && !url.isEmpty()) {
+            this.url = url;
```
[the remark this answers is not in the document]
Comment (leonardehrenfried) [reply src/main/java/org/opentripplanner/updater/car_park/ParkApiDataSource.java:68]: No, they don't. This is a straight up copy from the HSL updater.

File: src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java [hunk 4]
```diff
@@ -598,6 +598,26 @@ private static void addModeAndAlerts(Graph graph, Leg leg, State[] states, boole
...
+    }
+
+    private static void addCarParkAlerts(Leg leg, State[] state) {
+        var isTripPlannedForNow = Arrays.stream(state)
+                .findFirst()
+                .map(s -> s.getOptions().isTripPlannedForNow())
+                .orElse(false);
+
+        if(isTripPlannedForNow && containsCarParkWithFewSpaces(state)){
+            leg.addAlert(Alert.createLowCarParkSpacesAlert(), Locale.ENGLISH);
```
[the remark this answers is not in the document]
Comment (leonardehrenfried) [reply src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java]: As discussed, I changed it.

**Pull request merged**